sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -171,6 +173,57 @@ case class InSubqueryExec(
   }
 }
 
+/**
+ * The physical node of non-correlated EXISTS subquery.
+ */
+case class ExistsExec(
+    plan: BaseSubqueryExec,
+    exprId: ExprId)
+  extends ExecSubqueryExpression {
+
+  @transient private var result: Boolean = _
+
+  override def dataType: DataType = BooleanType
+  override def children: Seq[Expression] = Nil
+  override def nullable: Boolean = false
+  override def toString: String = s"EXISTS (${plan.simpleString(SQLConf.get.maxToStringFields)})"
+  override def withNewPlan(plan: BaseSubqueryExec): ExistsExec = copy(plan = plan)
+
+  override def semanticEquals(other: Expression): Boolean = other match {
+    case in: ExistsExec => plan.sameResult(in.plan)
+    case _ => false
+  }
+
+  def updateResult(): Unit = {
+    result = plan.executeTake(1).length == 1
+  }
+
+  def values(): Option[Boolean] = Option(result)
+
+  override def eval(input: InternalRow): Any = {
+    result
+  }
+
+  override lazy val canonicalized: ExistsExec = {
+    copy(
+      plan = plan.canonicalized.asInstanceOf[BaseSubqueryExec],
+      exprId = ExprId(0))
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
```

Comment:
again, can we follow `org.apache.spark.sql.execution.ScalarSubquery`? The codegen is well implemented there.

Reply:
as well as `eval`

Reply:
> again, can we follow `org.apache.spark.sql.execution.ScalarSubquery`? The codegen is well implemented there.

Done